src/Database/LSMTree/Internal.hs

```diff
@@ -723,29 +705,261 @@ updatesWithInterleavedFlushes conf hfs root uniqC es reg tc = do
+
+    -- Incoming runs being merged also need to be of the right size, but the
+    -- conditions are more complicated.
+    expectedMergingRunLengths ln mr mrs ls =
+      case mergePolicyForLevel (confMergePolicy conf) ln ls of
+        LevelLevelling ->
+          case (mr, mrs) of
+            -- A single incoming run (which thus didn't need merging) must be
+            -- of the expected size range already
+            (SingleRun r, CompletedMerge{}) -> assert (fits LevelLevelling r ln) True
+                        -- A completed merge for levelling can be of almost any size at all!
+            -- It can be smaller, due to deletions in the last level. But it
+            -- can't be bigger than would fit into the next level.
+            (_, CompletedMerge r) -> assert (fitsUB LevelLevelling r (ln + 1)) True
+        LevelTiering ->
+          case (mr, mrs, mergeLastForLevel ls) of
+            -- A single incoming run (which thus didn't need merging) must be
+            -- of the expected size already
+            (SingleRun r, CompletedMerge{}, _) -> assert (fits LevelTiering r ln) True
+
+            -- A completed last level run can be of almost any smaller size due
+            -- to deletions, but it can't be bigger than the next level down.
+            -- Note that tiering on the last level only occurs when there is
+            -- a single level only.
+            (_, CompletedMerge r, Merge.LastLevel) ->
+                assert (ln == 1) $
+                assert (fitsUB LevelTiering r (ln + 1)) $
+                True
+
+            -- A completed mid level run is usually of the size for the
+            -- level it is entering, but can also be one smaller (in which case
+            -- it'll be held back and merged again).
+            (_, CompletedMerge r, Merge.MidLevel) ->
+                assert (fitsUB LevelTiering r ln || fitsUB LevelTiering r (ln + 1)) True
+
+    -- Check that a run fits in the current level
+    fits policy r ln = fitsLB policy r ln && fitsUB policy r ln
+    -- Check that a run is too large for previous levels
+    fitsLB policy r ln = maxRunSize sr wba policy (ln - 1) < Run.runNumEntries r
+    -- Check that a run is too small for next levels
+    fitsUB policy r ln = Run.runNumEntries r <= maxRunSize sr wba policy ln
+
+{-# SPECIALISE addRunToLevels :: TableConfig -> HasFS IO h -> SessionRoot -> UniqCounter IO -> Run (Handle h) -> TempRegistry IO -> Levels (Handle h) -> IO (Levels (Handle h)) #-}
+addRunToLevels ::
+     forall m h. m ~ IO -- TODO: replace by @io-classes@ constraints for IO simulation.
+  => TableConfig
+  -> HasFS m h
+  -> SessionRoot
+  -> UniqCounter m
+  -> Run (Handle h)
+  -> TempRegistry m
+  -> Levels (Handle h)
+  -> m (Levels (Handle h))
+addRunToLevels conf@TableConfig{..} hfs root uniqC r0 reg levels = do
+    ls' <- go 1 (V.singleton r0) levels
+    () <- stToIO $ levelsInvariant conf ls'
```

Comment:
Oh, and checking the invariant needs to traverse the levels, even with asserts disabled. Since this is only done once per write buffer flush, it might be acceptable, but we could also make sure the whole traversal only happens on debug builds.